exchanges/okx/helpers.go

```diff
@@ -0,0 +1,214 @@
+	if instType != "" {
+		a, err := AssetTypeFromInstrumentType(instType)
+		if err != nil {
+			return nil, err
+		}
+		if a != asset.Empty {
+			return []asset.Item{a}, nil
+		}
+	}
+	if instrumentID == "" {
+		return nil, errMissingInstrumentID
+	}
+	pf, err := ok.CurrencyPairs.GetFormat(asset.Spot, true)
+	if err != nil {
+		return nil, err
+	}
+	splitSymbol := strings.Split(instrumentID, pf.Delimiter)
+	if len(splitSymbol) <= 1 {
+		return nil, fmt.Errorf("%w %v", currency.ErrCurrencyNotSupported, instrumentID)
+	}
+	pair, err := currency.NewPairDelimiter(instrumentID, pf.Delimiter)
+	if err != nil {
+		return nil, err
+	}
+	switch {
+	case len(splitSymbol) == 2:
+		resp := make([]asset.Item, 0, 2)
+		enabled, err := ok.IsPairEnabled(pair, asset.Spot)
+		if err != nil {
+			return nil, err
+		}
+		if enabled {
+			resp = append(resp, asset.Spot)
+		}
+		enabled, err = ok.IsPairEnabled(pair, asset.Margin)
+		if err != nil {
+			return nil, err
+		}
+		if enabled {
+			resp = append(resp, asset.Margin)
+		}
+		if len(resp) > 0 {
+			return resp, nil
+		}
+	case len(splitSymbol) > 2:
+		switch splitSymbol[len(splitSymbol)-1] {
+		case "SWAP", "swap":
+			enabled, err := ok.IsPairEnabled(pair, asset.PerpetualSwap)
+			if err != nil {
+				return nil, err
+			}
+			if enabled {
+				return []asset.Item{asset.PerpetualSwap}, nil
+			}
+		case "C", "P", "c", "p":
+			enabled, err := ok.IsPairEnabled(pair, asset.Options)
+			if err != nil {
+				return nil, err
+			}
+			if enabled {
+				return []asset.Item{asset.Options}, nil
+			}
+		default:
+			enabled, err := ok.IsPairEnabled(pair, asset.Futures)
+			if err != nil {
+				return nil, err
+			}
+			if enabled {
+				return []asset.Item{asset.Futures}, nil
+			}
+		}
+	}
+	return nil, fmt.Errorf("%w  '%v' or currency not enabled '%v'", asset.ErrNotSupported, instType, instrumentID)
+}
+
+// AssetTypeFromInstrumentType returns an asset Item instance given and Instrument Type string
+func AssetTypeFromInstrumentType(instrumentType string) (asset.Item, error) {
+	switch strings.ToUpper(instrumentType) {
+	case instTypeSwap, instTypeContract:
+		return asset.PerpetualSwap, nil
+	case instTypeSpot:
+		return asset.Spot, nil
+	case instTypeMargin:
+		return asset.Margin, nil
+	case instTypeFutures:
+		return asset.Futures, nil
+	case instTypeOption:
+		return asset.Options, nil
+	case "":
+		return asset.Empty, nil
+	default:
+		return asset.Empty, asset.ErrNotSupported
+	}
+}
+
+func (ok *Okx) validatePlaceOrderParams(arg *PlaceOrderRequestParam) error {
+	if *arg == (PlaceOrderRequestParam{}) {
+		return common.ErrEmptyParams
+	}
+	if arg.InstrumentID == "" {
+		return errMissingInstrumentID
+	}
+	arg.Side = strings.ToLower(arg.Side)
+	if arg.AssetType == asset.Spot || arg.AssetType == asset.Margin || arg.AssetType == asset.Empty {
+		arg.Side = strings.ToLower(arg.Side)
+		if arg.Side != order.Buy.Lower() && arg.Side != order.Sell.Lower() {
```

Comment:
Optional:
Could `PlaceOrderRequestParam` be an `order.Side` type, and stringify correctly?
A chunk of this would be simpler, and we'd just be saying, is it order.Buy or order.Sell 🙂 
Question probably applies to PositionSide as well.

Reply:
The reason for that was because when we use order.Side as a parameter and pass the object as an argument ... the JSON payload to be sent takes the number value than the  ("buy"/"sell") string value

Reply:
Just to leave a comment here; order.Side returns a Upper Case string of the type (eg. "SELL" or "BUY"). But Okx accepts the lowercase one ... If we have to use the default type _order.Side_ we have to implement an UnmarshaJSON() method for that with

Reply:
Acknowledged.

A dirty part of me is thinking about repurposing that bit shifting.
Could have 1-20 available as is using iota instead of 1 << iota
Then have bit masks for how to marshal it above that:
Uppercase, Lowercase, Capitalised = 3 bits, probably all the way to the right on a uint32.
Those could apply across a range of internal types.

Then it'd be like `order.Buy & types.Lowercase`
![image](https://github.com/user-attachments/assets/3279dad0-ce8c-4bf6-bab8-7e86d7eeb1fb)

Reply:
Hey @gbjk, I just saw this 😁. That sounds like a cool idea since we're not really using bit shifting for uniquely handling each element type. Let's get others' thoughts on it, and we can implement it based on your suggestion. This will help in validating the type and determining whether the string representation is lowercase, uppercase, or capitalized.

Reply:
I should have said: With the scope of this PR, this dirty thought is definitely out-of-scope.
Cute daydream though.